feat(inputs.firehose): Add new plugin

syedmhashim opened the pull request:
## Summary
<!-- Mandatory
Explain here the why, the rationale and motivation, for the changes.
-->
The firehose input plugin would be used to receive data from AWS Data Firehose.

## Checklist
<!-- Mandatory
Please confirm the following by replacing the space with an "x" between the []:
-->

- [x] No AI generated code was used in this PR

## Related issues
<!-- Mandatory
All PRs should resolve an issue, if one does not exist, please open one.
-->

resolves #15870

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+			if err != nil {
+				f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+			}
+			return
```
Please think about continuing parsing if that makes sense in your scenario to get as much information as we can...

syedmhashim replied:
I understand what you're saying but I do think we should return here because we are sending an `http.StatusBadRequest` response just above this code block so that firehose knows that the request was not able to go through and there are issues with the sent data (which I think is important)

srebhan replied:
You might want to add that as a comment to the code!

syedmhashim replied:
done

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+	}
+
+	commonAttributesHeaderValue := req.Header.Get(commonAttributesHeader)
+	if len(commonAttributesHeaderValue) != 0 && len(f.ParameterTags) != 0 {
```
Invert the logic and return early, this allows to save some nesting levels.
```suggestion
	if len(f.ParameterTags) == 0  || len(commonAttributesHeaderValue) == 0 {
		return
	}
```

syedmhashim replied:
We actually can't return here and would want to proceed to add the metrics to the accumulator. This block just adds tags to the metric if certain conditions are met. I'm unsure about moving this block to another function would be worth it. We can definitely return early from that function though. What do you think?

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+
+	commonAttributesHeaderValue := req.Header.Get(commonAttributesHeader)
+	if len(commonAttributesHeaderValue) != 0 && len(f.ParameterTags) != 0 {
+		parameters := make(map[string]interface{})
```
```suggestion
		var parameters map[string]interface{}
```

syedmhashim replied:
done

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+	commonAttributesHeaderValue := req.Header.Get(commonAttributesHeader)
+	if len(commonAttributesHeaderValue) != 0 && len(f.ParameterTags) != 0 {
+		parameters := make(map[string]interface{})
+		err := json.Unmarshal([]byte(commonAttributesHeaderValue), &parameters)
+		if err != nil {
```
```suggestion
		if err := json.Unmarshal([]byte(commonAttributesHeaderValue), &parameters); err != nil {
```

syedmhashim replied:
done

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+		})
+	}
+
+	commonAttributesHeaderValue := req.Header.Get(commonAttributesHeader)
```
Think about naming and try to be brief but expressive. How about `attributeHeaders` or `attributes`?

syedmhashim replied:
updated

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+
+		parameters, ok := parameters["commonAttributes"].(map[string]interface{})
+		if !ok {
+			f.Log.Warn(formatLog(r.body.RequestID, "Invalid value for header "+commonAttributesHeader))
```
Well I don't think you should simply continue here rather return here...

syedmhashim replied:
I thought of parameter tags as something optional. So if were able to extract and find them, we'd add to the metrics otherwise we'll ignore. Thus, I'm just emitting a warning log. Now that you mentioned, it does makes sense to maybe respond with bad request status and return here

srebhan replied:
Oh I see, but I would then at least skip the `for` loop below as this is known to not return anything but just burns some cycles...

syedmhashim replied:
updated

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+			f.Log.Warn(formatLog(r.body.RequestID, "Invalid value for header "+commonAttributesHeader))
+		}
+
+		for _, parameter := range f.ParameterTags {
+			if value, ok := parameters[parameter]; ok {
+				for _, m := range metrics {
+					m.AddTag(parameter, value.(string))
+				}
+			}
+		}
+	}
+
+	for _, m := range metrics {
+		if f.PathTag {
+			m.AddTag(pathTag, req.URL.Path)
+		}
+		f.acc.AddMetric(m)
+	}
```
I really would collect **all** tags _first_ and _then_ iterate over the metrics once to add them.

syedmhashim replied:
Okay. Does the following makes sense:
```
        if f.PathTag {
		for _, m := range metrics {
			m.AddTag(pathTag, req.URL.Path)
		}
	}
	
	for _, m := range metrics {
		f.acc.AddMetric(m)
	}
```

srebhan replied:
What I want is (pseudocode)
```golang
	tags := make(map[string]string, len(f.ParameterTags)+1)
	if f.pathTag {
		tags[f.pathTag] = req.URL.Path
	}
	for _, parameter := range f.ParameterTags {
		if value, found := parameters[parameter]; found {
			if v, ok := value.(string); ok {
				tags[parameter] = v
			}
		}
	}

	for _, m := range metrics {
		for k, v := range tags {
			m.AddTag(k , v)
		}
		f.acc.AddMetric(m)
	}
```

syedmhashim replied:
Actually `f.PathTag` variable is a boolean. When set to `True`, we add a tag, `firehose_http_path` to each metric with value set to the request path, i.e, `req.URL.Path`. It's unrelated to `f.ParameterTags`. The purpose of this was to allow the plugin to receive request on more than one request paths. So with the following configuration:
```toml
[[inputs.firehose]]
  ## Address and port to host HTTP listener on
  service_address = ":8080"

  ## Paths to listen to.
  paths = ["/telegraf/dev", "/telegraf/stage"]
 ```
our input plugin would accept requests for both of these paths. This would allow us to have a single telegraf server running, serving both our dev and stage environments. If we want, we can also add the prod environment here or have a separate telegraf server running for it. And if the `f.PathTag` is set to `True` we'll add a tag `firehose_http_path` to each metric with it's value set to the actual request path e.g `telegraf/dev`. This is something I thought could be useful and is optional therefore.

syedmhashim replied:
updated

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+	}
+}
+
+func formatLog(requestID, message string) string {
+	return fmt.Sprintf(logFormat, requestID, message)
+}
```
Please get rid of this function!

syedmhashim replied:
done

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+func init() {
+	inputs.Add("firehose", func() telegraf.Input {
+		return &Firehose{
+			ServiceAddress: ":8080",
+			TimeFunc:       time.Now,
+			Paths:          []string{"/telegraf"},
```
Please set the default values of `ServiceAddress` and `Paths` in `Init()` to reuse it in tests.

syedmhashim replied:
Makes sense. Btw if it's only for tests can't we do it the way it's done [here](https://github.com/influxdata/telegraf/blob/a16763497b89a65a536892e4ac8431506e3afc26/plugins/inputs/http_listener_v2/http_listener_v2_test.go#L52)? Just a thought

srebhan replied:
TBH, I hate these types of indirections. Imagine a test fails and you quickly want to look at what's going on. You immediately need to jump around in code to see how a simple test case is set up... :-( So please avoid this as possible!

syedmhashim replied:
done

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+			ServiceAddress: ":8080",
+			TimeFunc:       time.Now,
+			Paths:          []string{"/telegraf"},
+			close:          make(chan struct{}),
```
Also do in `Init()`.

srebhan reviewed (commented):
@syedmhashim I do have some comments, mostly on style, so overall this looks quite promising. I urge you to keep the code as simple and less nested as possible to ease review and debugging. And of course some unit-tests would be great!

syedmhashim commented:
> @syedmhashim I do have some comments, mostly on style, so overall this looks quite promising. I urge you to keep the code as simple and less nested as possible to ease review and debugging. And of course some unit-tests would be great!

Thanks for the review. Will definitely go over the comments and update accordingly. Just FYI, I have been using `http_listener_v2` input plugin as a reference to write the code. The unit tests will come soon as well

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+		if err != nil {
+			r.responseStatusCode = http.StatusBadRequest
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *firehoseRequest) decodeData() ([][]byte, bool) {
+	// decode base64-encoded data and return them as a slice of byte slices
+	decodedData := make([][]byte, 0)
+	for _, record := range r.body.Records {
+		data, err := base64.StdEncoding.DecodeString(record.EncodedData)
+		if err != nil {
+			return nil, false
+		}
+		decodedData = append(decodedData, data)
+	}
+	return decodedData, true
+}
+
+func (r *firehoseRequest) sendResponse(res http.ResponseWriter) error {
+	responseBody := struct {
+		RequestID    string `json:"requestId"`
+		Timestamp    int64  `json:"timestamp"`
+		ErrorMessage string `json:"errorMessage,omitempty"`
+	}{
+		RequestID:    r.req.Header.Get(requestIDHeader),
+		Timestamp:    time.Now().Unix(),
+		ErrorMessage: statusCodeToMessage[r.responseStatusCode],
+	}
+	response, err := json.Marshal(responseBody)
+	if err != nil {
+		return err
+	}
+	res.Header().Set("content-type", "application/json")
+	res.WriteHeader(r.responseStatusCode)
+	_, err = res.Write(response)
+	return err
+}
```
Good point! I'll have a look. If I'm not able to find something in the AWS SDK v2, then I'll definitely move this to a new file

syedmhashim replied:
done. moved to a separate file

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+//go:embed sample.conf
+var sampleConfig string
+
+var once sync.Once
```
Noted. Btw I do see this declared globally in every other (input) plugins

syedmhashim replied:
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+)
+
+// TimeFunc provides a timestamp for the metrics
+type TimeFunc func() time.Time
```
I saw this in `http_listener_v2` as well as in `influxdb_v2_listener`. Thought it's used somewhere downstream in telegraf. If this is not the case, please let me know and I'll remove this

srebhan replied:
Please remove. We had this in the dark age where we could not test metrics without pinning the timestamp...

syedmhashim replied:
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+		metrics = append(metrics, m...)
+	}
+
+	if len(metrics) == 0 {
+		once.Do(func() {
+			f.Log.Info(formatLog(r.body.RequestID, internal.NoMetricsCreatedMsg))
+		})
+	}
```
Yeah I guess so. Since no metrics created we can return. I'll add it. Thanks for pointing out. Again just copied over this block from `http_listener_v2` plugin

srebhan commented:
Yeah we do have some older code we didn't adapt yet but things changed both on the golang side as well as on us being more strict with the way things are done. ;-) No worries, we will figure it out together. :-)

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+	return nil
+}
+
+func (f *Firehose) ServeHTTP(res http.ResponseWriter, req *http.Request) {
```
Oh this is actually an implementation of the [http.Handler interface](https://pkg.go.dev/net/http#Handler). When we create the http server, we pass our plugin as the handler at line 262.

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+
+func (f *Firehose) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	select {
+	case <-f.close:
+		res.WriteHeader(http.StatusGone)
+		return
```
We do respond to the request using the `res.WriteHeader` method with `http.StatusGone` status code

srebhan replied:
Yes but you could also normally process the request instead of aborting it and exit afterwards (actually that's what should be done when shutting down the server). You do cover a very small time-range here which is a request was made just after we closed the server and just before reaching this code. If you are beyond this statement, the request will be answered anyway...

syedmhashim replied:
Well, I think then we can just get rid of this. Right?

syedmhashim replied:
removed

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+// Stop cleans up all resources
+func (f *Firehose) Stop() {
+	if f.listener != nil {
+		f.listener.Close()
```
Tbh, not sure. Again just copied this block from `http_listener_v2` plugin and does work as expected. I haven't yet inspected why this is used instead of `f.server.Shutdown`, but I'll definitely have a look.

srebhan replied:
Please take everything from http_listener_v2 with care as this is a _really_ old plugin. Check everything against current go standards!

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+
+	if f.AccessKey != "" {
+		if ok := r.authenticate(f.AccessKey); !ok {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), "Unauthorized"))
```
I understand what you're trying to say. The purpose of this was to associate the log message with the request ID so that we know for which specific request we are getting the log message. How do you propose I should format the log so that we also have an association with the request ID? An example would be really helpful

srebhan replied:
No I'm saying you should be explicit here and do
```golang
			f.Log.Error("unauthorized request %v from %v", req.Header.Get(requestIDHeader), r.RemoteAddr)
```
as we immediately can find the error message in the code.

syedmhashim replied:
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+	pathTag   = "firehose_http_path"
+	logFormat = "RequestID:%s Message:%s"
+
+	// request headers
+	requestIDHeader        = "x-amz-firehose-request-id"
+	accessKeyHeader        = "x-amz-firehose-access-key"
+	commonAttributesHeader = "x-amz-firehose-common-attributes"
```
Noted. Do you want me to do the same for the following?
```
// defaultMaxBodySize is the default maximum request body size, in bytes.
// if the request body is over this size, we will return an HTTP 413 error.
// 64 MB
const defaultMaxBodySize = 64 * 1024 * 1024

const (
	pathTag   = "firehose_http_path"
)
```

srebhan replied:
Yeah, including `pathTag` and `defaultMaxBodySize`. There is no value in defining those constants here as they will speak for themselves in the code at the places where you use them and we don't need to jump around in the code when debugging/reviewing.

syedmhashim replied:
done

srebhan commented:
@syedmhashim I'm loosing a bit the track in the discussion above, could you please push an update with the stuff that is clear and then we discuss the unclear parts?

syedmhashim commented on plugins/inputs/firehose/README.md:
```diff
@@ -0,0 +1,123 @@
+# AWS Data Firehose HTTP Listener Input Plugin
+
+Firehose HTTP Listener is a service input plugin that listens for metrics sent
```
done

syedmhashim commented on plugins/inputs/firehose/README.md:
```diff
@@ -0,0 +1,123 @@
+    [[inputs.firehose.json_v2.tag]]
+      path = "type"
+    [[inputs.firehose.json_v2.field]]
+      path = "message"
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+
+	listener net.Listener
+
+	telegraf.Parser
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+
+// Start starts the http listener service.
+func (f *Firehose) Start(acc telegraf.Accumulator) error {
+	if f.ReadTimeout < config.Duration(time.Second) {
+		f.ReadTimeout = config.Duration(time.Second * 10)
+	}
+	if f.WriteTimeout < config.Duration(time.Second) {
+		f.WriteTimeout = config.Duration(time.Second * 10)
+	}
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+		defer f.wg.Done()
+		if err := server.Serve(f.listener); err != nil {
+			if !errors.Is(err, net.ErrClosed) {
+				f.Log.Errorf("Serve failed - %s", err.Error())
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+	}
+	if err != nil {
+		return err
+	}
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+	PathTag        bool            `toml:"path_tag"`
+	ReadTimeout    config.Duration `toml:"read_timeout"`
+	WriteTimeout   config.Duration `toml:"write_timeout"`
+	AccessKey      string          `toml:"access_key"`
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+
+	if f.AccessKey != "" {
+		if ok := r.authenticate(f.AccessKey); !ok {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), "Unauthorized"))
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+		WriteTimeout: time.Duration(f.WriteTimeout),
+		TLSConfig:    f.tlsConf,
+	}
+}
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+	if err := r.validate(); err != nil {
+		f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		err := r.sendResponse(res)
+		if err != nil {
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+	if !ok {
+		f.Log.Error(formatLog(r.body.RequestID, "Failed to base64 decode record data"))
+		err := r.sendResponse(res)
+		if err != nil {
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,413 @@
+		return
+	}
+
+	metrics := make([]telegraf.Metric, 0)
```
done

syedmhashim commented:
@srebhan Hey! Just pushed some changes and commented "done" on the threads that I resolved. Sorry for the delay. I got occupied with my job

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+	if err := r.authenticate(f.AccessKey); err != nil {
+		f.Log.Error(err.Error())
+		if err = r.sendResponse(res); err != nil {
+			f.Log.Errorf("error sending response to request %s, %v", requestID, err.Error())
```
Please use a consistent error pattern like
```suggestion
			f.Log.Errorf("sending response to request %q failed: %v", requestID, err)
```

I.e. `"<action> <text with params> failed: %v", params..., err` as this allows to easily see what went wrong and has the error after the colon...

Furthermore, you don't need `err.Error()` with the `%v` directive, `err` is sufficient.

Same for all error logs below!

syedmhashim replied:
done

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+		return
+	}
+
+	decodedBytesData, ok := r.decodeData()
```
Please be brief with your variable names! Developers are lazy. ;-)
```suggestion
	data, ok := r.decodeData()
```

syedmhashim replied:
done

srebhan commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,156 @@
+	responseStatusCode int
+}
+
+func (r *firehoseRequest) authenticate(expectedAccessKey config.Secret) error {
```
```suggestion
func (r *firehoseRequest) authenticate(expected config.Secret) error {
```

syedmhashim replied:
done

srebhan commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,156 @@
+	reqAccessKey := r.req.Header.Get("x-amz-firehose-access-key")
+	if reqAccessKey != accessKey.String() {
+		r.responseStatusCode = http.StatusUnauthorized
+		return fmt.Errorf("unauthorized request %s", r.req.Header.Get("x-amz-firehose-request-id"))
```
Please log the source of the request (aka remote address) if possible.

syedmhashim replied:
done

srebhan commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,156 @@
+		return errors.New("x-amz-firehose-request-id header is not set")
+	}
+
+	// Check if content length is not over 64 MB.
```
Is this a documented limit? If so, please put a reference in this comment!

syedmhashim replied:
done

srebhan commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,156 @@
+		return fmt.Errorf("unaccepted content type, %s, in request %s", contentType, requestID)
+	}
+
+	contentEncoding := r.req.Header.Get("content-encoding")
+	if contentEncoding != "" && contentEncoding != "gzip" {
+		r.responseStatusCode = http.StatusBadRequest
+		return fmt.Errorf("unaccepted content encoding, %s, in request %s", contentEncoding, requestID)
+	}
+
+	err := r.extractBody()
+	if err != nil {
+		return err
+	}
```
I would combine those and not move it to an own function as it hampers the reading flow
```suggestion
	encoding := r.req.Header.Get("content-encoding")
	body, err := internal.NewStreamContentDecoder(encoding, r.req.Body)
	if err != nil {
		return fmt.Errorf("creating decoder for %q failed: %w", encoding, err)
	}
	defer r.req.Body.Close()
	
	if err := json.NewDecoder(body).Decode(&r.body); err != nil {
		r.responseStatusCode = http.StatusBadRequest
		return fmt.Errorf("decode gzip body failed: %w", err)
	}
```

syedmhashim replied:
Noted. Atm, firehose can only be configured to use gzip as per the [documentation](https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html). Do you think it'd be good to keep the gzip check which I already added?
```
if contentEncoding != "" && contentEncoding != "gzip" {
		r.responseStatusCode = http.StatusBadRequest
		return fmt.Errorf("unaccepted content encoding, %s, in request %s", contentEncoding, requestID)
	}
```

syedmhashim replied:
Nvm. Figured we only use the gzip decoder in the `NewStreamContentDecoder` method

syedmhashim replied:
done

srebhan commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+func (r *firehoseRequest) sendResponse(res http.ResponseWriter) error {
+	responseBody := struct {
+		RequestID    string `json:"requestId"`
+		Timestamp    int64  `json:"timestamp"`
+		ErrorMessage string `json:"errorMessage,omitempty"`
+	}{
```
Please declare this structure as you do with the request etc.

syedmhashim replied:
done

srebhan reviewed (commented):
That looks much better @syedmhashim! Some more cleanups and comments from my side...

syedmhashim commented on plugins/inputs/firehose/README.md:
```diff
@@ -0,0 +1,117 @@
+  # path_tag = false
+
+  ## maximum duration before timing out read of the request
+  # read_timeout = "10s"
+  ## maximum duration before timing out write of the response
+  # write_timeout = "10s"
```
what about 5 seconds?

syedmhashim replied:
udpated to 5 secs

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+		return
+	}
+
+	r := &firehoseRequest{req: req}
+	requestID := req.Header.Get("x-amz-firehose-request-id")
+
```
We do that in the `validate` method. Do we want to do it again here?
```
func (r *firehoseRequest) validate() error {
	requestID := r.req.Header.Get("x-amz-firehose-request-id")
	if requestID == "" {
		r.responseStatusCode = http.StatusBadRequest
		return errors.New("x-amz-firehose-request-id header is not set")
	}
        \\ more validations
}
```

syedmhashim replied:
Maybe we can move the validation above authentication and everything else so that the first thing we check is if this header is set

syedmhashim replied:
Done

syedmhashim commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+func (r *firehoseRequest) authenticate(expectedAccessKey config.Secret) error {
+	if expectedAccessKey.Empty() {
+		return nil
+	}
```
If the expected key is empty, it means that there is no authentication configured on telegraf and thus we don't expect any key. Therefore, even if a key is provided in the incoming data it doesn't matter

srebhan replied:
Please reflect this in the `sample.conf` comment and also make this very clear in the README! This is somewhat unexpected to me as it means "accept all" which is a really bad security strategy. ;-)

syedmhashim replied:
I had kept authentication optional so that if the `access_key` is not set in the configuration block it means accept all. Otherwise, we’d check the incoming request for access key header and compare it with the one in the configuration block. However, we can also update the plugin to make the access key as something required and not optional. What do you suggest?

syedmhashim commented on plugins/inputs/firehose/README.md:
```diff
@@ -0,0 +1,117 @@
+
+  ## Save path as firehose_http_path tag if set to true
+  # path_tag = false
+
```
done(removed this)

syedmhashim commented on plugins/inputs/firehose/README.md:
```diff
@@ -0,0 +1,117 @@
+  ## Each data format has its own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
+  data_format = "influx"
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+	http.StatusRequestEntityTooLarge: "request body too large",
+	http.StatusUnauthorized:          "unauthorized",
+	http.StatusOK:                    "",
+}
```
done

syedmhashim commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,156 @@
+	if !isAcceptedMethod {
+		r.responseStatusCode = http.StatusMethodNotAllowed
+		return fmt.Errorf("forbidden method, %s, in request %s", r.req.Method, requestID)
+	}
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+	}
+	if f.WriteTimeout < config.Duration(time.Second) {
+		f.WriteTimeout = config.Duration(time.Second * 10)
+	}
```
I did not get this :/  Can you share any existing example

syedmhashim replied:
Need more context on this

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+			if !errors.Is(err, net.ErrClosed) {
+				f.Log.Errorf("starting server failed: %v", err)
+			}
+			close(f.close)
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+}
+
+func (f *Firehose) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	if !choice.Contains(req.URL.Path, f.Paths) {
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+
+// Stop cleans up all resources
+func (f *Firehose) Stop() {
+	if f.listener != nil {
```
done(no longer applicable)

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+	requestID := req.Header.Get("x-amz-firehose-request-id")
+
+	if err := r.authenticate(f.AccessKey); err != nil {
+		f.Log.Error(err.Error())
```
done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+	for _, bytes := range decodedBytesData {
+		m, err := f.parser.Parse(bytes)
+		if err != nil {
+			f.Log.Errorf("unable to parse data from request %s", requestID)
```
done

syedmhashim commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,156 @@
+	req                *http.Request
+	body               firehoseRequestBody
+	responseStatusCode int
+}
```
done

syedmhashim commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,156 @@
+		return fmt.Errorf("getting accesskey failed: %w", err)
+	}
+	reqAccessKey := r.req.Header.Get("x-amz-firehose-access-key")
+	if reqAccessKey != accessKey.String() {
```
done

syedmhashim commented:
@srebhan Hi! Pushed some changes and added comments where relevant

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+			f.Log.Warnf("x-amz-firehose-common-attributes header's value is not a valid json in request %s", requestID)
+		}
+
+		parameters, ok := parameters["commonAttributes"].(map[string]interface{})
```
Added a new method, `addParmaterTags`, to Firehose struct

srebhan commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,140 @@
+	// https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html
+	if r.req.ContentLength > int64(64*1024*1024) {
+		r.res.statusCode = http.StatusRequestEntityTooLarge
+		return fmt.Errorf("content length too large in request %s", requestID)
```
I would add the request ID when printing the error of `validate`, i.e. one caller level up and remove it from the error messages here. What do you think? It's not a must and mostly cosmetic...

srebhan commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+func (r *request) authenticate(expected config.Secret) error {
+	if expected.Empty() {
+		return nil
+	}
```
Please put a comment here
```suggestion
	// We completely switch off authentication if not 'access_key' was provided in the config, it's intended!
	if expected.Empty() {
		return nil
	}
```

srebhan commented on plugins/inputs/firehose/sample.conf:
```diff
@@ -0,0 +1,40 @@
+  ## Minimal TLS version accepted by the server
+  # tls_min_version = "TLS12"
+
+  ## Optional access key to accept for authentication.
+  ## AWS Data Firehose uses "x-amz-firehose-access-key" header to set the access key
```
```suggestion
  ## Optional access key to accept for authentication.
  ## AWS Data Firehose uses "x-amz-firehose-access-key" header to set the access key.
  ## If no access_key is provided (default), authentication is completely disabled and
  ## this plugin will accept all request ignoring the provided access-key in the request!
  
```

srebhan commented on plugins/inputs/firehose/firehose_request.go:
```diff
@@ -0,0 +1,140 @@
+	key := r.req.Header.Get("x-amz-firehose-access-key")
+	match, err := expected.EqualTo([]byte(key))
+	if err != nil {
+		return fmt.Errorf("comparing keys failed: %w", err)
```
You don't set a status code here! What happens with the response to the sender? Shouldn't this be a 500? Maybe you should make sure that in `sendResponse` (or when creating the request) you will set a defined fallback status code (maybe 500?)!

syedmhashim replied:
Oh I missed that. And agreed with the fallback status code

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+	}
+
+	var metrics []telegraf.Metric
+	for _, bytes := range data {
```
Please don't use `bytes` as variable name as this is also the name of a built-in Go library. How about `sample` , `msg` or `record`?

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,241 @@
+		return
+	}
+
+	if err := f.addParameterTags(r, metrics); err != nil {
+		f.Log.Errorf("add parameter tags for request %q failed: %v", requestID, err)
+		r.res.statusCode = http.StatusBadRequest
+		if err = r.sendResponse(res); err != nil {
+			f.Log.Errorf("sending response to request %q failed: %v", requestID, err)
+		}
+		return
+	}
+
+	for _, m := range metrics {
+		m.AddTag("firehose_http_path", req.URL.Path)
+	}
+
+	for _, m := range metrics {
+		f.acc.AddMetric(m)
+	}
```
How about
```suggestion
	tags, err := r.extractParameterTags(f.ParameterTags)
	if err != nil {
		f.Log.Errorf("extracting parameter tags for request %q failed: %v", requestID, err)
		r.res.statusCode = http.StatusBadRequest
		if err = r.sendResponse(res); err != nil {
			f.Log.Errorf("sending response to request %q failed: %v", requestID, err)
		}
		return
	}

	for _, m := range metrics {
		for k, v := range tags {
			m.AddTag(k, v)
		}
		m.AddTag("firehose_http_path", req.URL.Path)
		f.acc.AddMetric(m)
	}
```

This way you don't need to iterate over the metrics three times and do have a clear code structure where everything related to processing the request is in the request "class"...

syedmhashim replied:
makes sense

srebhan reviewed (commented):
@syedmhashim thanks for the update. I tried to put more concrete suggestions into my review. Overall, I think we are almost there, the only thing missing are unit-tests with a mocked sender...

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,216 @@
+func init() {
+	inputs.Add("firehose", func() telegraf.Input {
+		return &Firehose{
+			ServiceAddress: ":8080",
+			Paths:          []string{"/telegraf"},
```
Sure. Just a question. What's the difference between initialising the attributes in `init()`and in `Init()` and how do we decide which option to choose?

srebhan replied:
The `init()` (lowercase) function will be called when importing the package by Go itself. So what you do here is how the structure is registered and that's how each instance looks like **before parsing any config options**! The `Init()` (uppercase) function is called **after setting the struct fields from the config options**. So you _should_ do everything in `Init()` if the `nil` value is invalid for the type and option (like strings etc). However, if you do have for example timeouts where a zero-value might mean "wait forever" it is better to set this in `init()` (lowercase) as the user might want to overwrite the default with zero. Does that make sense?

syedmhashim replied:
Yes, makes sense.

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,216 @@
+		return err
+	}
+
+	return nil
```
Done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,216 @@
+		}
+	}()
+
+	f.Log.Infof("listening on %s", f.listener.Addr().String())
```
Done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,216 @@
+	err := f.server.Shutdown(context.Background())
+	if err != nil {
+		f.Log.Errorf("shutting down server failed: %v", err)
+	}
```
Done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,216 @@
+		r.res.statusCode = http.StatusBadRequest
+		f.Log.Errorf("x-amz-firehose-request-id header is not set")
+		if err := r.sendResponse(res); err != nil {
+			f.Log.Errorf("sending response failed: %v", err)
```
Done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,216 @@
+		return
+	}
+
+	r := &request{req: req}
```
Done

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,216 @@
+	paramTags, err := r.extractParameterTags(f.ParameterTags)
+	if err != nil {
+		f.Log.Errorf("extracting parameter tags for request %q failed: %v", requestID, err)
+		r.res.statusCode = http.StatusBadRequest
```
Done

syedmhashim commented:
Hi @srebhan. In addition to the changes that you suggested, I also made some changes to the code(see last commit) that I think would make it more clean and readable. P.S I think it also helps to achieve separation of concern since now all processing of the request is handled by `firehose_request.go` while `firehose.go` mostly focus on adding metrics:

Changes to :
- `firehose_request.go`
    - Added `newRequest` method to handle the creation of request object and (gzip +) json decoding of the request body
    - Added `processRequest` method to handle the authentication, validation, base64 decoding and parameter extraction
    - I’ve tried to set the response status code mostly in `firehose_request.go`
- `firehose.go`
    - Added `handleRequest` method. This helps to get rid of the redundant `sendResponse` code block.

I'd like to hear your thoughts on this

srebhan commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,178 @@
+	}
+}
+
+func (f *Firehose) handleRequest(req *http.Request) (r *request) {
+	var err error
+	if r, err = newRequest(req); err != nil {
+		f.Log.Errorf("Creating request object failed: %v", err)
+		return r
+	}
```
How about returning the actual error here and do the logging in `ServeHTTP`? This will save a few lines I think
```suggestion
func (f *Firehose) handleRequest(req *http.Request) (*request, error) {
	r, err := newRequest(req)
	if err != nil {
		return fmt.Errorf("creating request failed: %w", err)
	}
```
of course you need to modify the code below and in `ServeHTTP`...

syedmhashim replied:
I tried implementing this but doesn’t help. The line count remains the same as we need to introduce err checking in the `ServeHTTP` function. I made some other changes which saves us a couple lines though

srebhan reviewed (commented):
@syedmhashim thanks for the update. I do have some smaller comments in the code but nothing big.

Regarding the tests, I suggest that you don't test the request separately but simply use a http client and send data to the plugin. You might also want to implement a general test-case setup similar to what we do in [socket listener tests](https://github.com/influxdata/telegraf/blob/0ea4c1422ef013b39e10061ce96ac4bdb0ccd93e/plugins/inputs/socket_listener/socket_listener_test.go#L351)...

If you need help with the tests, please provide some payload examples and invite me to your branch and I'll help you to drive this PR over the finish line...

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,178 @@
+		f.listener, err = net.Listen("tcp", f.ServiceAddress)
+	}
+	if err != nil {
+		return err
```
updated

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,178 @@
+	go func() {
+		if err := f.server.Serve(f.listener); err != nil {
+			if !errors.Is(err, net.ErrClosed) {
+				f.Log.Errorf("Starting server failed: %v", err)
```
Updated

syedmhashim commented on plugins/inputs/firehose/firehose.go:
```diff
@@ -0,0 +1,178 @@
+// Stop cleans up all resources
+func (f *Firehose) Stop() {
+	err := f.server.Shutdown(context.Background())
+	if err != nil {
```
updated

syedmhashim commented:
@srebhan Hey. Just sent you an invite! The payload needs to be in the following format with `data` having the actual b64 encoded text that would be parsed (based on the parser set) after decoding:
```
{
    "requestId": “test-id",
    "timestamp": 123456789,
    "records": [
        {
          "data": "aGVsbG8gd29ybGQK" // "hello world"
        }
    ]
}
```
